### PR TITLE
Report errors that occur in the project file, in the IDE as well

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Changes
- Report errors that occur in the project file, in the IDE as well
- Show a hint in the project file that shows which files the project references

### Testing
- Added two unit tests

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyDriver/Commands/CommandRegistry.cs [hunk 1]
```diff
@@ -270,10 +270,20 @@ class WritersConsole : IConsole {
...
         return false;
       }
-      var projectFile = await DafnyProject.Open(OnDiskFileSystem.Instance, new Uri(singleFile.FullName), dafnyOptions.OutputWriter, dafnyOptions.ErrorWriter);
+      var projectFile = await DafnyProject.Open(OnDiskFileSystem.Instance, new Uri(singleFile.FullName));
       if (projectFile == null) {
         return false;
       }
+
+      foreach (var diagnostic in projectFile.Errors.AllMessages) {
+        var message = diagnostic.Level + ": " + diagnostic.Message;
```
Comment (MikaelMayer) [inline Source/DafnyDriver/Commands/CommandRegistry.cs:279]:
```suggestion
        var message = $"{diagnostic.Level}: {diagnostic.Message}";
```

File: Source/DafnyCore/Options/DafnyProject.cs [hunk 2]
```diff
@@ -19,37 +19,56 @@ public class DafnyProject : IEquatable<DafnyProject> {
...
-      await errorWriter.WriteLineAsync($"The Dafny project file {uri.LocalPath} contains the following errors:");
       var regex = new Regex(
         @$"\((\d+),(\d+)\) : error : The property `(\w+)` was not found on object type {typeof(DafnyProject).FullName}");
       var newMessage = regex.Replace(tomlException.Message,
         match =>
           $"({match.Groups[1].Value},{match.Groups[2].Value}): the property {match.Groups[3].Value} does not exist.");
-      await errorWriter.WriteLineAsync(newMessage);
-      return null;
+      result = emptyProject;
+      result.Errors.Error(MessageSource.Parser, result.StartingToken, $"The Dafny project file {uri.LocalPath} contains the following errors: " + newMessage);
```
Comment (MikaelMayer) [inline Source/DafnyCore/Options/DafnyProject.cs:64]:
```suggestion
      result.Errors.Error(MessageSource.Parser, result.StartingToken, $"The Dafny project file {uri.LocalPath} contains the following errors: {newMessage}");
```

**Review by MikaelMayer — COMMENTED** [review]
That sounds great to have. Can you please add a docs/dev/news entry if appropriate?

File: Source/DafnyLanguageServer.Test/Util/ClientBasedLanguageServerTest.cs [hunk 3]
```diff
@@ -158,6 +158,39 @@ protected ClientBasedLanguageServerTest(ITestOutputHelper output, LogLevel dafny
...
+
+    foreach (var entry in Projects.Managers) {
+      try {
+        await entry.GetLastDocumentAsync().WaitAsync(cancellationToken.Value);
+      } catch (TaskCanceledException) {
+
+      }
+    }
+
+    var verificationDocumentItem = CreateTestDocument("class X {does not parse", $"AssertNoDiagnosticsAreComing{fileIndex++}.dfy");
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer.Test/Util/ClientBasedLanguageServerTest.cs:172]: I love this non-parsing file :-)

File: Source/DafnyLanguageServer.Test/Util/ClientBasedLanguageServerTest.cs [hunk 4]
```diff
@@ -158,6 +158,39 @@ protected ClientBasedLanguageServerTest(ITestOutputHelper output, LogLevel dafny
...
+      }
+
+      result = result.Add(resolutionReport.Uri, resolutionReport.Diagnostics.ToList());
+    }
+    client.DidCloseTextDocument(new DidCloseTextDocumentParams {
+      TextDocument = verificationDocumentItem
+    });
+    var hideReport = await diagnosticsReceiver.AwaitNextNotificationAsync(cancellationToken.Value);
+    AssertM.Equal(verificationDocumentItem.Uri, hideReport.Uri,
+      "2) Unexpected diagnostics were received whereas none were expected:\n" +
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer.Test/Util/ClientBasedLanguageServerTest.cs:189]: Why is `2)` there?

**Review by MikaelMayer — APPROVED** [review]
One non-blocking comment. It's great that we can now debug the project file from the IDE. If I change the project file, I assumed it's picked up again as soon as I modify the file, correct?

**Comment by keyboardDrummer** [comment]
> One non-blocking comment. It's great that we can now debug the project file from the IDE. If I change the project file, I assumed it's picked up again as soon as I modify the file, correct?

That requires a change in the VSCode extension, because currently the extension only sends `.dfy` file `didOpen` and `didChange` event to the server, so for the `dfyconfig.toml` only the on-disk state is used by the server.

I plan to make that change but it is independent from this one.

File: Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs [hunk 5]
```diff
@@ -78,7 +78,7 @@ ILogger<ITextDocumentLoader> logger
         var projectDirectory = Path.GetDirectoryName(projectPath)!;
         var filesMessage = string.Join("\n", compilation.RootUris.Select(uri => Path.GetRelativePath(projectDirectory, uri.LocalPath)));
         if (filesMessage.Any()) {
-          program.Reporter.Info(MessageSource.Parser, compilation.Project.StartingToken, "Files referenced by project are:\n" + filesMessage);
+          program.Reporter.Info(MessageSource.Parser, compilation.Project.StartingToken, "Files referenced by project are:" + Environment.NewLine + filesMessage);
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs:81]: Oh yeah thanks for making sure Windows tests don't break!

**Review by MikaelMayer — APPROVED** [review]
Thanks a lot!

**Pull request merged**